kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionBuilderImpl.java

```diff
@@ -373,24 +373,26 @@ private void validateMetadataChange(
         throw new KernelException("Cannot update mapping mode and perform schema evolution");
       }
 
+      // If the column mapping restriction is removed, clustering columns
+      // will need special handling during schema evolution since they won't have physical names
+      // ToDo: Support adding clustering columns
+
       if (!isColumnMappingModeEnabled(updatedMappingMode)) {
         throw new KernelException("Cannot update schema for table when column mapping is disabled");
       }
 
-      // TODO: revisit this once we want to support schema evolution with clustering columns
-      Optional<List<Column>> clusteringColumns =
-          ClusteringUtils.getClusteringColumnsOptional(snapshot);
-      if (clusteringColumns.isPresent() && !clusteringColumns.get().isEmpty()) {
-        throw new KernelException(
-            format(
-                "Update schema for table with clustering columns %s is not yet supported",
-                clusteringColumns.get()));
-      }
+      // Clustering columns will be guaranteed to have physical names at this point
+      Set<String> clusteringColumnPhysicalNames =
+          ClusteringUtils.getClusteringColumnsOptional(snapshot).orElse(Collections.emptyList())
+              .stream()
+              .map(col -> col.getNames()[0])
```

Comment:
why do we only want the top-level name?

Reply:
I realized it should actually be only the leaf field when testing this a bit more. The getClusteringColumn API is a bit odd in that it returns a `Column` where every part is just the physical name so like `<physical_coluumn_for_struct>.<physical_column_for_inner_field>. `

I feel like the right thing to do is just return the leaf field for that API. There's no use as far as I can tell to the other parts when column mapping is enabled.


Anyways, since validation is performed on every leaf field, we just need to the leaf field physical name, that uniquely identifies the inner field anyways.

Reply:
I'll add a comment for this since I realize it's not obvious

Reply:
@allisonport-db let me know if the comment makes sense, I'll leave this thread unresolved for now